gamemodes/terrortown/entities/entities/ttt_smoke.lua

```diff
@@ -0,0 +1,428 @@
+		print("picked color", color_name, "or", color_dex, "out of", #colors_available)
+		color_override = self.SmokeColors[ color_name ]
+	end
+
+	local mix = math.Clamp(RandomWrap(
+		(self:GetColorBase() - self:GetColorVariance()) / 255,
+		(self:GetColorBase() + self:GetColorVariance()) / 255
+	), 0, 1)
+
+	if color_override ~= nil then
+		local h, s = ColorToHSV( color_override )
+		color_override = HSVToColor(h, s, mix)
+	else
+		color_override = HSVToColor(0, 0, mix)
+		-- legacy:
+		-- local unmix = math.Round(mix*255)
+		-- Color(unmix, unmix, unmix)
+	end
+
+	return color_override
+end
+
+---
+-- Gets a point within a sphere that is suitably random, and uses our PRNG (SharedRandom).
+-- @todo move to a library or util, idk
+-- @realm shared
+function ENT:PointInSphere()
+	local RandomWrap = self.RandomWrap
+
+	local u = RandomWrap(0, 1)
+	local v = RandomWrap(0, 1)
+	local theta = u * 2.0 * math.pi
+	local phi = math.acos(2.0 * v - 1.0)
+	local r = math.pow(RandomWrap(0, 1), 1 / 2)
+	local sinTheta = math.sin(theta)
+	local cosTheta = math.cos(theta)
+	local sinPhi = math.sin(phi)
+	local cosPhi = math.cos(phi)
+	local x = r * sinPhi * cosTheta
+	local y = r * sinPhi * sinTheta
+	local z = r * cosPhi
+	return Vector(x, y, z)
+end
+
+---
+-- Initiates the smoke effect and spawns all the particles, by reading all the data on this entity.
+-- @realm client
+function ENT:ClientEffect()
+	local RandomWrap = self.RandomWrap
+	self.Emitter = ParticleEmitter(self:GetPos(), false)
+	self.Emitter:SetPos(self:GetPos())
+	local c = self:GetSmokeColor():ToColor()
+	local diameter = self:GetRadius() * 2
+	print("particles for:", self:GetRadius(), "d:", diameter, self:GetStartSizeBase(), "+/-",  self:GetStartSizeVariance(), "legend:")
+	print("particle:\tstart_size\tmax_range\toffset_ratio\toffset_dimension")
+	for i = 1, self:GetDensity() do
+		--TODO: raycast points
+		local start_size = math.Clamp(RandomWrap(
+			self:GetStartSizeBase() - self:GetStartSizeVariance(),
+			self:GetStartSizeBase() + self:GetStartSizeVariance()
+		), 1, diameter)
+		local max_range = diameter - (start_size / 2)
+		local offset_ratio = (max_range / diameter)
+		local offset_dimension = self:GetRadius() * offset_ratio
+		local rel_pos = self:PointInSphere() * offset_dimension
+		print(
+			string.format("particle:\t%.6f\t%.6f\t%.6f\t%.6f",  start_size, max_range, offset_ratio, offset_dimension)
+		)
+
+		local sprite_dex = math.max(1,math.ceil(RandomWrap(0, #self.SmokeParticles)))
+		local p = self.Emitter:Add(self.SmokeParticles[sprite_dex], self:GetPos() + rel_pos)
+
+		self.Particles[ #self.Particles + 1 ] = p
+
+		if p then
+			p:SetColor(c.r, c.g, c.b)
+			p:SetStartAlpha(255)
+			p:SetEndAlpha(255)
+
+			-- we don't need to set position?
+			-- p:SetPos(self:GetPos() + rel_pos)
+			p:SetVelocity((p:GetPos() - self:GetPos()):GetNormal() * self.DriftVelocity)
+
+			p:SetLifeTime(0)
+			local die_time = self:GetLifeTime() + self:GetLifeTime()
+			p:SetDieTime(die_time)
```

Comment:
this looks weird. Does this do anything? SetLifeTime is set to 0, which is probably important. But the die time is then also 0?

Reply:
`self` and `p` are different, what i was doing here was a test to see if there was something I could tweak to prevent the Emitter from failing to render on the client. it didn't work, so this gets gutted.